# 📊 Notebook 04: Product & Division Analysis

## Product Line Profitability & Margin Performance Analysis  
### Nassau Candy Distributor

---

## 🔷 Objective

The objective of this notebook is to analyze product-level and division-level profitability performance.

This notebook focuses on:

- Product-level profitability aggregation
- Product segment performance
- Top and bottom product identification
- Division-level revenue and profit comparison
- Business insights and recommendations

### Import Libraries

In [4]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)
pd.set_option("display.float_format", "{:,.2f}".format)

print("Libraries imported successfully.")

Libraries imported successfully.


## 🔷 1. Load Feature Engineered Dataset

We load the feature engineered dataset created in Notebook 03.

In [5]:
file_path = "../data/processed/feature_engineered_nassau_candy.csv"

df = pd.read_csv(file_path)

print("Feature engineered dataset loaded successfully.")
print("Shape:", df.shape)

Feature engineered dataset loaded successfully.
Shape: (10194, 24)


## Quick Check

In [6]:
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Country/Region,City,State/Province,Postal Code,Division,Region,Product ID,Product Name,Sales,Units,Gross Profit,Cost,Gross Margin %,Profit per Unit,Revenue Contribution %,Profit Contribution %,Order Month,Product Segment
0,1,US-2021-103800-CHO-MIL-31000,2024-01-03,2026-06-30,Standard Class,103800,United States,Houston,Texas,77095,Chocolate,Interior,CHO-MIL-31000,Wonka Bar - Milk Chocolate,6.50,2,4.22,2.28,0.65,2.11,0.00,0.00,2024-01,Underperformers
1,2,US-2021-112326-CHO-TRI-54000,2024-01-04,2026-07-01,Standard Class,112326,United States,Naperville,Illinois,60540,Chocolate,Interior,CHO-TRI-54000,Wonka Bar - Triple Dazzle Caramel,7.50,2,4.90,2.60,0.65,2.45,0.00,0.00,2024-01,Volume Traps
2,3,US-2021-112326-CHO-NUT-13000,2024-01-04,2026-07-01,Standard Class,112326,United States,Naperville,Illinois,60540,Chocolate,Interior,CHO-NUT-13000,Wonka Bar - Nutty Crunch Surprise,10.47,3,7.47,3.00,0.71,2.49,0.00,0.00,2024-01,Stars
3,4,US-2021-112326-CHO-SCR-58000,2024-01-04,2026-07-01,Standard Class,112326,United States,Naperville,Illinois,60540,Chocolate,Interior,CHO-SCR-58000,Wonka Bar -Scrumdiddlyumptious,10.80,3,7.50,3.30,0.69,2.50,0.00,0.00,2024-01,Stars
4,5,US-2021-141817-CHO-TRI-54000,2024-01-05,2026-07-05,Standard Class,141817,United States,Philadelphia,Pennsylvania,19143,Chocolate,Atlantic,CHO-TRI-54000,Wonka Bar - Triple Dazzle Caramel,11.25,3,7.35,3.90,0.65,2.45,0.00,0.00,2024-01,Underperformers


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10194 entries, 0 to 10193
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Row ID                  10194 non-null  int64  
 1   Order ID                10194 non-null  object 
 2   Order Date              10194 non-null  object 
 3   Ship Date               10194 non-null  object 
 4   Ship Mode               10194 non-null  object 
 5   Customer ID             10194 non-null  int64  
 6   Country/Region          10194 non-null  object 
 7   City                    10194 non-null  object 
 8   State/Province          10194 non-null  object 
 9   Postal Code             10194 non-null  object 
 10  Division                10194 non-null  object 
 11  Region                  10194 non-null  object 
 12  Product ID              10194 non-null  object 
 13  Product Name            10194 non-null  object 
 14  Sales                   10194 non-null

## 🔷 2. Product-Level Aggregation

In this step, we aggregate data at the product level to evaluate overall performance.

We calculate:

- Total Sales
- Total Profit
- Total Units
- Average Margin
- Average Profit per Unit

In [7]:
# Product-level aggregation

product_df = df.groupby("Product Name").agg({
    "Sales": "sum",
    "Cost": "sum",   # 👈 ADD THIS
    "Gross Profit": "sum",
    "Units": "sum",
    "Gross Margin %": "mean",
    "Profit per Unit": "mean"
}).reset_index()

product_df.head()

,Product Name,Sales,Cost,Gross Profit,Units,Gross Margin %,Profit per Unit
0,Everlasting Gobstopper,130.00,26.00,104.00,13,0.80,8.00
1,Fizzy Lifting Drinks,78.75,31.50,47.25,21,0.60,2.25
2,Fun Dip,12.00,7.20,4.80,8,0.40,0.60
3,Hair Toffee,76.50,17.00,59.50,17,0.78,3.50
4,Kazookles,"1,205.75","1,113.00",92.75,371,0.08,0.25


### Add Contribution Metrics (Product-Level)

In [8]:
# Recalculate contributions at product level

total_sales = product_df["Sales"].sum()
total_profit = product_df["Gross Profit"].sum()

product_df["Revenue Contribution %"] = product_df["Sales"] / total_sales
product_df["Profit Contribution %"] = product_df["Gross Profit"] / total_profit

### 📌 Verify

In [9]:
product_df.head(15)

,Product Name,Sales,Cost,Gross Profit,Units,Gross Margin %,Profit per Unit,Revenue Contribution %,Profit Contribution %
0,Everlasting Gobstopper,130.00,26.00,104.00,13,0.80,8.00,0.00,0.00
1,Fizzy Lifting Drinks,78.75,31.50,47.25,21,0.60,2.25,0.00,0.00
2,Fun Dip,12.00,7.20,4.80,8,0.40,0.60,0.00,0.00
3,Hair Toffee,76.50,17.00,59.50,17,0.78,3.50,0.00,0.00
4,Kazookles,"1,205.75","1,113.00",92.75,371,0.08,0.25,0.01,0.00
5,Laffy Taffy,53.73,20.25,33.48,27,0.62,1.24,0.00,0.00
6,Lickable Wallpaper,"7,860.00","3,930.00","3,930.00",393,0.50,10.00,0.06,0.04
7,Nerds,15.00,8.00,7.00,10,0.47,0.70,0.00,0.00
8,SweeTARTS,61.50,32.80,28.70,41,0.47,0.70,0.00,0.00
9,Wonka Bar - Fudge Mallows,"24,890.40","8,296.80","16,593.60",6914,0.67,2.40,0.18,0.18


## 🔷 3. Product Ranking

We rank products based on:

- Total Sales
- Total Profit

This helps identify:
- Top performing products
- Low performing products

### Top by Sales

In [10]:
# Top products by sales

product_df.sort_values(by="Sales", ascending=False).head(5)

,Product Name,Sales,Cost,Gross Profit,Units,Gross Margin %,Profit per Unit,Revenue Contribution %,Profit Contribution %
12,Wonka Bar - Triple Dazzle Caramel,"28,485.00","9,874.80","18,610.20",7596,0.65,2.45,0.20,0.20
13,Wonka Bar -Scrumdiddlyumptious,"27,874.80","8,517.30","19,357.50",7743,0.69,2.50,0.20,0.21
10,Wonka Bar - Milk Chocolate,"26,867.75","9,424.38","17,443.37",8267,0.65,2.11,0.19,0.19
9,Wonka Bar - Fudge Mallows,"24,890.40","8,296.80","16,593.60",6914,0.67,2.40,0.18,0.18
11,Wonka Bar - Nutty Crunch Surprise,"23,574.95","6,755.00","16,819.95",6755,0.71,2.49,0.17,0.18


### Top by Profit

In [11]:
# Top products by profit

product_df.sort_values(by="Gross Profit", ascending=False).head(5)

,Product Name,Sales,Cost,Gross Profit,Units,Gross Margin %,Profit per Unit,Revenue Contribution %,Profit Contribution %
13,Wonka Bar -Scrumdiddlyumptious,"27,874.80","8,517.30","19,357.50",7743,0.69,2.50,0.20,0.21
12,Wonka Bar - Triple Dazzle Caramel,"28,485.00","9,874.80","18,610.20",7596,0.65,2.45,0.20,0.20
10,Wonka Bar - Milk Chocolate,"26,867.75","9,424.38","17,443.37",8267,0.65,2.11,0.19,0.19
11,Wonka Bar - Nutty Crunch Surprise,"23,574.95","6,755.00","16,819.95",6755,0.71,2.49,0.17,0.18
9,Wonka Bar - Fudge Mallows,"24,890.40","8,296.80","16,593.60",6914,0.67,2.40,0.18,0.18


### Bottom Products

In [12]:
# Bottom products by profit

product_df.sort_values(by="Gross Profit", ascending=True).head(5)

,Product Name,Sales,Cost,Gross Profit,Units,Gross Margin %,Profit per Unit,Revenue Contribution %,Profit Contribution %
2,Fun Dip,12.00,7.20,4.80,8,0.40,0.60,0.00,0.00
7,Nerds,15.00,8.00,7.00,10,0.47,0.70,0.00,0.00
8,SweeTARTS,61.50,32.80,28.70,41,0.47,0.70,0.00,0.00
5,Laffy Taffy,53.73,20.25,33.48,27,0.62,1.24,0.00,0.00
1,Fizzy Lifting Drinks,78.75,31.50,47.25,21,0.60,2.25,0.00,0.00


### 🔷 Top Performing Products (By Sales)

The highest revenue-generating products are:

- Wonka Bar - Triple Dazzle Caramel (~28.4K sales)
- Wonka Bar - Scrumdiddlyumptious (~27.8K sales)
- Wonka Bar - Milk Chocolate (~26.8K sales)

These products dominate overall revenue contribution and indicate strong market demand.

### 🔷 Top Performing Products (By Profit)

The most profitable products are:

- Wonka Bar - Scrumdiddlyumptious (~19.3K profit)
- Wonka Bar - Triple Dazzle Caramel (~18.6K profit)
- Wonka Bar - Milk Chocolate (~17.4K profit)

These products not only generate high revenue but also maintain strong margins, making them key drivers of profitability.

### 🔷 Low Performing Products (Bottom by Profit)

The least profitable products are:

- Fun Dip (~4.8 profit)
- Nerds (~7.0 profit)
- SweeTARTS (~28.7 profit)

These products contribute very little to overall profit and may require:

- Pricing optimization
- Cost reduction
- Strategic repositioning

## 🔷 4. Segment-Level Performance Analysis

In this step, we analyze product performance based on predefined segments.

Segments:

- ⭐ Stars → High Profit + High Margin
- 💎 Hidden Gems → High Profit + Low Margin
- ⚠️ Volume Traps → High Sales but Low Profit
- ❌ Underperformers → Low Sales + Low Profit

This analysis helps identify:

- Where the company is making money
- Where inefficiencies exist
- Which products require action

### Segment Distribution

In [13]:
df["Product Segment"].value_counts()

Product Segment
Stars              3881
Underperformers    3128
Hidden Gems        1818
Volume Traps       1367
Name: count, dtype: int64

### Segment-Level Aggregation

In [14]:
segment_df = df.groupby("Product Segment").agg({
    "Sales": "sum",
    "Gross Profit": "sum",
    "Units": "sum",
    "Gross Margin %": "mean",
    "Profit per Unit": "mean"
}).reset_index()

segment_df

,Product Segment,Sales,Gross Profit,Units,Gross Margin %,Profit per Unit
0,Hidden Gems,"24,890.40","16,593.60",6914,0.67,2.40
1,Stars,"51,656.25","36,340.95",14528,0.70,2.50
2,Underperformers,"40,764.48","25,724.75",12389,0.63,2.07
3,Volume Traps,"24,472.50","14,783.50",4823,0.64,2.97


### Add Contribution Metrics

In [15]:
total_sales = segment_df["Sales"].sum()
total_profit = segment_df["Gross Profit"].sum()

segment_df["Revenue Contribution %"] = segment_df["Sales"] / total_sales
segment_df["Profit Contribution %"] = segment_df["Gross Profit"] / total_profit

segment_df

,Product Segment,Sales,Gross Profit,Units,Gross Margin %,Profit per Unit,Revenue Contribution %,Profit Contribution %
0,Hidden Gems,"24,890.40","16,593.60",6914,0.67,2.40,0.18,0.18
1,Stars,"51,656.25","36,340.95",14528,0.70,2.50,0.36,0.39
2,Underperformers,"40,764.48","25,724.75",12389,0.63,2.07,0.29,0.28
3,Volume Traps,"24,472.50","14,783.50",4823,0.64,2.97,0.17,0.16


### 🔷 Segment Performance Insights

⭐ Stars:
These products contribute the highest share of both revenue and profit, making them the backbone of the business. These should be prioritized for scaling and promotion.

💎 Hidden Gems:
These products generate strong profit but operate at lower margins. There may be pricing inefficiencies, indicating an opportunity to improve margins.

⚠️ Volume Traps:
These products generate high sales but low profit, indicating cost-heavy or inefficient pricing strategies. These require urgent optimization.

❌ Underperformers:
These products contribute minimal revenue and profit. They may need discontinuation or repositioning.

### 🔷 Final Segment Interpretation

⭐ Stars dominate both revenue and profit, indicating strong product-market fit and efficient pricing.

💎 Hidden Gems show high profitability but relatively lower margin efficiency, suggesting potential for margin optimization.

⚠️ Volume Traps generate significant sales but low profit, indicating cost inefficiencies or poor pricing strategies. These products require urgent optimization.

❌ Underperformers contribute the least to both revenue and profit, making them the weakest segment. These products should be evaluated for discontinuation or repositioning.

## 🔷 5. Division-Level Performance Analysis

In this step, we analyze profitability performance across product divisions.

This helps identify:

- Which division generates the highest revenue
- Which division contributes the most profit
- Which division has stronger margin efficiency
- Whether any division shows revenue-profit imbalance

### Division-Level Aggregation

In [16]:
division_df = df.groupby("Division").agg({
    "Sales": "sum",
    "Gross Profit": "sum",
    "Cost": "sum",
    "Units": "sum",
    "Gross Margin %": "mean",
    "Profit per Unit": "mean"
}).reset_index()

division_df

,Division,Sales,Gross Profit,Cost,Units,Gross Margin %,Profit per Unit
0,Chocolate,"131,692.90","88,824.62","42,868.28",37275,0.67,2.38
1,Other,"9,663.25","4,333.45","5,329.80",1242,0.38,3.36
2,Sugar,427.48,284.73,142.75,137,0.58,1.89


### Add Contribution Metrics

In [17]:
total_sales = division_df["Sales"].sum()
total_profit = division_df["Gross Profit"].sum()

division_df["Revenue Contribution %"] = division_df["Sales"] / total_sales
division_df["Profit Contribution %"] = division_df["Gross Profit"] / total_profit

division_df

,Division,Sales,Gross Profit,Cost,Units,Gross Margin %,Profit per Unit,Revenue Contribution %,Profit Contribution %
0,Chocolate,"131,692.90","88,824.62","42,868.28",37275,0.67,2.38,0.93,0.95
1,Other,"9,663.25","4,333.45","5,329.80",1242,0.38,3.36,0.07,0.05
2,Sugar,427.48,284.73,142.75,137,0.58,1.89,0.00,0.00


### Sort by Profit

In [18]:
division_df.sort_values(by="Gross Profit", ascending=False)

,Division,Sales,Gross Profit,Cost,Units,Gross Margin %,Profit per Unit,Revenue Contribution %,Profit Contribution %
0,Chocolate,"131,692.90","88,824.62","42,868.28",37275,0.67,2.38,0.93,0.95
1,Other,"9,663.25","4,333.45","5,329.80",1242,0.38,3.36,0.07,0.05
2,Sugar,427.48,284.73,142.75,137,0.58,1.89,0.00,0.00


### 🔷 Division-Level Insights

🍫 Chocolate Division:
This division dominates both revenue (~93%) and profit (~95%), making it the primary driver of overall business performance. It shows strong scalability and should be the core focus for growth and expansion.

📦 Other Division:
Although this division contributes a smaller share of revenue (~7%), it has the highest profit per unit, indicating strong margin efficiency. This suggests premium or high-margin products that could be scaled strategically.

🍬 Sugar Division:
This division contributes negligible revenue and profit, making it the weakest performer. It may require strategic review, including pricing optimization, cost control, or potential discontinuation of low-performing products.

⚠️ Key Business Insight:
The business is highly dependent on the Chocolate division, indicating a concentration risk. Diversification across other divisions is recommended to reduce dependency and improve overall stability.

## 🔷 6. Cost vs Profit Diagnostics

In this step, we analyze the relationship between cost and profit.

This helps identify:

- Cost-heavy products
- Low-margin products
- Pricing inefficiencies
- Products that require optimization

### Cost vs Sales Relationship

In [19]:
df[["Cost", "Sales", "Gross Profit"]].describe()

,Cost,Sales,Gross Profit
count,"10,194.00","10,194.00","10,194.00"
mean,4.74,13.91,9.17
std,5.06,11.34,6.64
min,0.60,1.25,0.25
25%,2.40,7.20,4.90
50%,3.60,10.80,7.47
75%,5.70,18.00,12.25
max,130.00,260.00,130.00


### Product-Level Cost Analysis

In [20]:
product_cost_df = df.groupby("Product Name").agg({
    "Sales": "sum",
    "Cost": "sum",
    "Gross Profit": "sum"
}).reset_index()

product_cost_df["Cost Ratio"] = product_cost_df["Cost"] / product_cost_df["Sales"]

product_cost_df.head()

,Product Name,Sales,Cost,Gross Profit,Cost Ratio
0,Everlasting Gobstopper,130.00,26.00,104.00,0.20
1,Fizzy Lifting Drinks,78.75,31.50,47.25,0.40
2,Fun Dip,12.00,7.20,4.80,0.60
3,Hair Toffee,76.50,17.00,59.50,0.22
4,Kazookles,"1,205.75","1,113.00",92.75,0.92


### Identify High Cost Products

In [21]:
product_cost_df.sort_values(by="Cost Ratio", ascending=False).head(5)

,Product Name,Sales,Cost,Gross Profit,Cost Ratio
4,Kazookles,"1,205.75","1,113.00",92.75,0.92
2,Fun Dip,12.00,7.20,4.80,0.60
7,Nerds,15.00,8.00,7.00,0.53
8,SweeTARTS,61.50,32.80,28.70,0.53
6,Lickable Wallpaper,"7,860.00","3,930.00","3,930.00",0.50


### Identify Low Margin Products

In [22]:
product_df.sort_values(by="Gross Margin %").head(5)

,Product Name,Sales,Cost,Gross Profit,Units,Gross Margin %,Profit per Unit,Revenue Contribution %,Profit Contribution %
4,Kazookles,"1,205.75","1,113.00",92.75,371,0.08,0.25,0.01,0.00
2,Fun Dip,12.00,7.20,4.80,8,0.40,0.60,0.00,0.00
7,Nerds,15.00,8.00,7.00,10,0.47,0.70,0.00,0.00
8,SweeTARTS,61.50,32.80,28.70,41,0.47,0.70,0.00,0.00
6,Lickable Wallpaper,"7,860.00","3,930.00","3,930.00",393,0.50,10.00,0.06,0.04


### 🔷 Cost vs Profit Insights

⚠️ High Cost Ratio Products:
Products with high cost-to-sales ratio indicate inefficient cost structures or pricing issues. These products consume a large portion of revenue, reducing profitability.

❌ Low Margin Products:
Products with low gross margin indicate weak pricing power or high production costs. These require pricing revision or cost optimization.

💡 Business Recommendation:
- Optimize pricing for low-margin products
- Reduce cost through supplier negotiation
- Consider discontinuation of consistently low-margin products

### 🔷 Cost vs Profit Insight (Final)

Kazookles is identified as the most cost-heavy product and also the lowest margin product.

Despite generating relatively high sales volume, its cost-to-sales ratio is extremely high, resulting in minimal profit contribution. This indicates a clear pricing inefficiency or high production cost structure.

⚠️ Business Risk:
Kazookles is a classic "Volume Trap" product — it drives revenue but does not translate into profitability.

💡 Recommended Actions:
- Re-evaluate pricing strategy to improve margins
- Investigate cost structure for potential reduction
- Consider repositioning or discontinuation if inefficiencies persist

## 🔷 7. Pareto Analysis (80/20 Rule)

In this step, we identify the concentration of revenue and profit across products.

This helps determine:

- Which products drive the majority of revenue
- Which products drive the majority of profit
- Whether the business is dependent on a few key products

### Sort by Revenue

In [23]:
pareto_sales = product_df.sort_values(by="Sales", ascending=False).copy()

pareto_sales["Cumulative Sales %"] = pareto_sales["Sales"].cumsum() / pareto_sales["Sales"].sum()

pareto_sales

,Product Name,Sales,Cost,Gross Profit,Units,Gross Margin %,Profit per Unit,Revenue Contribution %,Profit Contribution %,Cumulative Sales %
12,Wonka Bar - Triple Dazzle Caramel,"28,485.00","9,874.80","18,610.20",7596,0.65,2.45,0.20,0.20,0.20
13,Wonka Bar -Scrumdiddlyumptious,"27,874.80","8,517.30","19,357.50",7743,0.69,2.50,0.20,0.21,0.40
10,Wonka Bar - Milk Chocolate,"26,867.75","9,424.38","17,443.37",8267,0.65,2.11,0.19,0.19,0.59
9,Wonka Bar - Fudge Mallows,"24,890.40","8,296.80","16,593.60",6914,0.67,2.40,0.18,0.18,0.76
11,Wonka Bar - Nutty Crunch Surprise,"23,574.95","6,755.00","16,819.95",6755,0.71,2.49,0.17,0.18,0.93
6,Lickable Wallpaper,"7,860.00","3,930.00","3,930.00",393,0.50,10.00,0.06,0.04,0.98
4,Kazookles,"1,205.75","1,113.00",92.75,371,0.08,0.25,0.01,0.00,0.99
14,Wonka Gum,597.50,286.80,310.70,478,0.52,0.65,0.00,0.00,1.00
0,Everlasting Gobstopper,130.00,26.00,104.00,13,0.80,8.00,0.00,0.00,1.00
1,Fizzy Lifting Drinks,78.75,31.50,47.25,21,0.60,2.25,0.00,0.00,1.00


### Identify 80% Revenue Products

In [24]:
top_80_sales = pareto_sales[pareto_sales["Cumulative Sales %"] <= 0.80]

top_80_sales

,Product Name,Sales,Cost,Gross Profit,Units,Gross Margin %,Profit per Unit,Revenue Contribution %,Profit Contribution %,Cumulative Sales %
12,Wonka Bar - Triple Dazzle Caramel,"28,485.00","9,874.80","18,610.20",7596,0.65,2.45,0.20,0.20,0.20
13,Wonka Bar -Scrumdiddlyumptious,"27,874.80","8,517.30","19,357.50",7743,0.69,2.50,0.20,0.21,0.40
10,Wonka Bar - Milk Chocolate,"26,867.75","9,424.38","17,443.37",8267,0.65,2.11,0.19,0.19,0.59
9,Wonka Bar - Fudge Mallows,"24,890.40","8,296.80","16,593.60",6914,0.67,2.40,0.18,0.18,0.76


### Sort by Profit

In [25]:
pareto_profit = product_df.sort_values(by="Gross Profit", ascending=False).copy()

pareto_profit["Cumulative Profit %"] = pareto_profit["Gross Profit"].cumsum() / pareto_profit["Gross Profit"].sum()

pareto_profit

,Product Name,Sales,Cost,Gross Profit,Units,Gross Margin %,Profit per Unit,Revenue Contribution %,Profit Contribution %,Cumulative Profit %
13,Wonka Bar -Scrumdiddlyumptious,"27,874.80","8,517.30","19,357.50",7743,0.69,2.50,0.20,0.21,0.21
12,Wonka Bar - Triple Dazzle Caramel,"28,485.00","9,874.80","18,610.20",7596,0.65,2.45,0.20,0.20,0.41
10,Wonka Bar - Milk Chocolate,"26,867.75","9,424.38","17,443.37",8267,0.65,2.11,0.19,0.19,0.59
11,Wonka Bar - Nutty Crunch Surprise,"23,574.95","6,755.00","16,819.95",6755,0.71,2.49,0.17,0.18,0.77
9,Wonka Bar - Fudge Mallows,"24,890.40","8,296.80","16,593.60",6914,0.67,2.40,0.18,0.18,0.95
6,Lickable Wallpaper,"7,860.00","3,930.00","3,930.00",393,0.50,10.00,0.06,0.04,0.99
14,Wonka Gum,597.50,286.80,310.70,478,0.52,0.65,0.00,0.00,1.00
0,Everlasting Gobstopper,130.00,26.00,104.00,13,0.80,8.00,0.00,0.00,1.00
4,Kazookles,"1,205.75","1,113.00",92.75,371,0.08,0.25,0.01,0.00,1.00
3,Hair Toffee,76.50,17.00,59.50,17,0.78,3.50,0.00,0.00,1.00


### Identify 80% Profit Products

In [26]:
top_80_profit = pareto_profit[pareto_profit["Cumulative Profit %"] <= 0.80]

top_80_profit

,Product Name,Sales,Cost,Gross Profit,Units,Gross Margin %,Profit per Unit,Revenue Contribution %,Profit Contribution %,Cumulative Profit %
13,Wonka Bar -Scrumdiddlyumptious,"27,874.80","8,517.30","19,357.50",7743,0.69,2.50,0.20,0.21,0.21
12,Wonka Bar - Triple Dazzle Caramel,"28,485.00","9,874.80","18,610.20",7596,0.65,2.45,0.20,0.20,0.41
10,Wonka Bar - Milk Chocolate,"26,867.75","9,424.38","17,443.37",8267,0.65,2.11,0.19,0.19,0.59
11,Wonka Bar - Nutty Crunch Surprise,"23,574.95","6,755.00","16,819.95",6755,0.71,2.49,0.17,0.18,0.77


### 🔷 Pareto Insights (80/20 Rule)

Revenue Concentration:
A small number of products contribute to the majority of total revenue, indicating a concentration of business performance around key products.

Profit Concentration:
Similarly, a limited set of products generates the majority of total profit, highlighting the most valuable products for the company.

⚠️ Business Risk:
The company is heavily dependent on a few products for both revenue and profit, creating a dependency risk.

💡 Recommendations:
- Strengthen top-performing products (Stars)
- Reduce dependency by improving performance of other products
- Optimize underperforming products to balance portfolio

### 🔷 Pareto Analysis Insights (80/20 Rule)

Revenue Concentration:
Approximately 80% of total revenue is generated by just **2 products**, indicating a strong concentration of sales within a very small subset of the product portfolio.

Profit Concentration:
Even more critically, **80% of total profit is driven by only 1 product**, highlighting extreme dependency on a single high-performing product.

⚠️ Business Risk:
This indicates a **high concentration risk**, where the business is heavily dependent on a limited number of products—especially one key product—for profitability.

Any disruption in demand, pricing, or supply of this product could significantly impact overall business performance.

💡 Strategic Recommendations:
- Strengthen and protect top-performing products (especially the top profit driver)
- Reduce dependency by improving performance of mid-tier and underperforming products
- Focus on converting "Hidden Gems" into "Stars"
- Re-evaluate low-margin products (Volume Traps) for cost optimization or discontinuation

## 🔷 8. Export Analysis Datasets

In this step, we export all analysis-ready datasets generated in Notebook 04.

These files will be saved inside the `data/processed` folder and will be used for dashboard development, reporting, and final business storytelling.

Exported datasets include:

- Final feature-engineered dataset
- Product-level analysis
- Division-level analysis
- Top sales products
- Top profit products
- Bottom products
- Segment-level performance
- Cost vs profit diagnostics
- Pareto sales analysis
- Pareto profit analysis

In [27]:
# =========================================
# EXPORT ANALYSIS DATASETS
# =========================================

import os

# Create processed folder if it does not exist
os.makedirs("../data/processed", exist_ok=True)

# -----------------------------------------
# 1. Main datasets
# -----------------------------------------

df.to_csv("../data/processed/final_feature_engineered_dataset.csv", index=False)
product_df.to_csv("../data/processed/product_level_analysis.csv", index=False)
division_df.to_csv("../data/processed/division_level_analysis.csv", index=False)

# -----------------------------------------
# 2. Product ranking datasets
# -----------------------------------------

top_sales = product_df.sort_values(by="Sales", ascending=False)
top_profit = product_df.sort_values(by="Gross Profit", ascending=False)
bottom_products = product_df.sort_values(by="Gross Profit", ascending=True)

top_sales.to_csv("../data/processed/top_sales_products.csv", index=False)
top_profit.to_csv("../data/processed/top_profit_products.csv", index=False)
bottom_products.to_csv("../data/processed/bottom_products.csv", index=False)

# -----------------------------------------
# 3. Segment-level performance dataset
# -----------------------------------------

segment_summary = df.groupby("Product Segment").agg({
    "Sales": "sum",
    "Cost": "sum",
    "Gross Profit": "sum",
    "Units": "sum",
    "Gross Margin %": "mean",
    "Profit per Unit": "mean"
}).reset_index()

segment_summary["Revenue Contribution %"] = segment_summary["Sales"] / segment_summary["Sales"].sum()
segment_summary["Profit Contribution %"] = segment_summary["Gross Profit"] / segment_summary["Gross Profit"].sum()

segment_summary.to_csv("../data/processed/segment_performance.csv", index=False)

# -----------------------------------------
# 4. Cost vs profit diagnostics dataset
# -----------------------------------------

cost_profit_df = product_df.copy()

cost_profit_df["Cost Ratio"] = cost_profit_df["Cost"] / cost_profit_df["Sales"]

cost_profit_df = cost_profit_df.sort_values(by="Cost Ratio", ascending=False)

cost_profit_df.to_csv("../data/processed/cost_vs_profit_analysis.csv", index=False)

# -----------------------------------------
# 5. Pareto analysis datasets
# -----------------------------------------

pareto_sales.to_csv("../data/processed/pareto_sales_analysis.csv", index=False)
pareto_profit.to_csv("../data/processed/pareto_profit_analysis.csv", index=False)

# -----------------------------------------
# Final confirmation
# -----------------------------------------

print("✅ All Notebook 04 analysis datasets exported successfully to data/processed/")

✅ All Notebook 04 analysis datasets exported successfully to data/processed/
